// @okxweb3/x402-core/dist/esm/chunk-E363UOMT.mjs
// src/schemas/index.ts
import { z } from "zod";
import { z as z2 } from "zod";
var NonEmptyString = z.string().min(1);
var Any = z.record(z.unknown());
var OptionalAny = z.record(z.unknown()).optional().nullable();
var NetworkSchema = z.string().min(3).refine((val) => val.includes(":"), {
  message: "Network must be in CAIP-2 format (e.g., 'eip155:84532')"
});
var ResourceInfoSchema = z.object({
  url: NonEmptyString,
  description: z.string().optional(),
  mimeType: z.string().optional()
});
var PaymentRequirementsSchema = z.object({
  scheme: NonEmptyString,
  network: NetworkSchema,
  amount: NonEmptyString,
  asset: NonEmptyString,
  payTo: NonEmptyString,
  maxTimeoutSeconds: z.number().positive(),
  extra: OptionalAny
});
var PaymentRequiredSchema = z.object({
  x402Version: z.literal(2),
  error: z.string().optional(),
  resource: ResourceInfoSchema,
  accepts: z.array(PaymentRequirementsSchema).min(1),
  extensions: OptionalAny
});
var PaymentPayloadSchema = z.object({
  x402Version: z.literal(2),
  resource: ResourceInfoSchema.optional(),
  accepted: PaymentRequirementsSchema,
  payload: Any,
  extensions: OptionalAny
});
function parsePaymentRequired(value) {
  return PaymentRequiredSchema.safeParse(value);
}
function validatePaymentRequired(value) {
  return PaymentRequiredSchema.parse(value);
}
function isPaymentRequired(value) {
  return PaymentRequiredSchema.safeParse(value).success;
}
function parsePaymentRequirements(value) {
  return PaymentRequirementsSchema.safeParse(value);
}
function validatePaymentRequirements(value) {
  return PaymentRequirementsSchema.parse(value);
}
function isPaymentRequirements(value) {
  return PaymentRequirementsSchema.safeParse(value).success;
}
function parsePaymentPayload(value) {
  return PaymentPayloadSchema.safeParse(value);
}
function validatePaymentPayload(value) {
  return PaymentPayloadSchema.parse(value);
}
function isPaymentPayload(value) {
  return PaymentPayloadSchema.safeParse(value).success;
}

export {
  NonEmptyString,
  Any,
  OptionalAny,
  NetworkSchema,
  ResourceInfoSchema,
  PaymentRequirementsSchema,
  PaymentRequiredSchema,
  PaymentPayloadSchema,
  parsePaymentRequired,
  validatePaymentRequired,
  isPaymentRequired,
  parsePaymentRequirements,
  validatePaymentRequirements,
  isPaymentRequirements,
  parsePaymentPayload,
  validatePaymentPayload,
  isPaymentPayload,
  z2 as z
};
