// @okxweb3/x402-core/dist/esm/chunk-O3IYMTNT.mjs
// src/facilitator/OKXFacilitatorClient.ts
import crypto from "crypto";
var OKXFacilitatorClient = class {
  /**
   *
   * @param config
   */
  constructor(config) {
    this.config = {
      baseUrl: "https://web3.okx.com",
      ...config
    };
  }
  /**
   *
   * @param method
   * @param path
   * @param body
   */
  createHeaders(method, path, body) {
    const timestamp = (/* @__PURE__ */ new Date()).toISOString();
    const prehash = timestamp + method + path + (body ?? "");
    const sign = crypto.createHmac("sha256", this.config.secretKey).update(prehash).digest("base64");
    return {
      "OK-ACCESS-KEY": this.config.apiKey,
      "OK-ACCESS-SIGN": sign,
      "OK-ACCESS-TIMESTAMP": timestamp,
      "OK-ACCESS-PASSPHRASE": this.config.passphrase,
      "Content-Type": "application/json"
    };
  }
  /**
   *
   */
  async getSupported() {
    const path = "/api/v6/pay/x402/supported";
    const res = await fetch(this.config.baseUrl + path, {
      headers: this.createHeaders("GET", path)
    });
    if (!res.ok) throw new Error(`OKX getSupported failed: ${res.status}`);
    const json = await res.json();
    const data = json.data ?? json;
    return data;
  }
  /**
   *
   * @param payload
   * @param requirements
   */
  async verify(payload, requirements) {
    const path = "/api/v6/pay/x402/verify";
    const body = JSON.stringify({
      x402Version: 2,
      paymentPayload: payload,
      paymentRequirements: requirements
    });
    const res = await fetch(this.config.baseUrl + path, {
      method: "POST",
      headers: this.createHeaders("POST", path, body),
      body
    });
    if (!res.ok) throw new Error(`OKX verify failed: ${res.status}`);
    const json = await res.json();
    const data = json.data ?? json;
    return data;
  }
  /**
   *
   * @param payload
   * @param requirements
   */
  async settle(payload, requirements) {
    const path = "/api/v6/pay/x402/settle";
    const bodyObj = {
      x402Version: 2,
      paymentPayload: payload,
      paymentRequirements: requirements
    };
    if (this.config.syncSettle !== void 0) {
      bodyObj.syncSettle = this.config.syncSettle;
    }
    const body = JSON.stringify(bodyObj);
    const res = await fetch(this.config.baseUrl + path, {
      method: "POST",
      headers: this.createHeaders("POST", path, body),
      body
    });
    if (!res.ok) throw new Error(`OKX settle failed: ${res.status}`);
    const json = await res.json();
    const data = json.data ?? json;
    return data;
  }
  /**
   * Query on-chain settlement status by transaction hash.
   *
   * @param txHash - The transaction hash to query
   * @returns Settlement status response
   */
  async getSettleStatus(txHash) {
    const path = `/api/v6/pay/x402/settle/status?txHash=${encodeURIComponent(txHash)}`;
    const res = await fetch(this.config.baseUrl + path, {
      headers: this.createHeaders("GET", path)
    });
    if (!res.ok) throw new Error(`OKX getSettleStatus failed: ${res.status}`);
    const json = await res.json();
    const data = json.data ?? json;
    return data;
  }
};

// src/index.ts
var x402Version = 2;

export {
  OKXFacilitatorClient,
  x402Version
};
