// @okxweb3/x402-core/dist/cjs/schemas/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/schemas/index.ts
var schemas_exports = {};
__export(schemas_exports, {
  Any: () => Any,
  NetworkSchema: () => NetworkSchema,
  NonEmptyString: () => NonEmptyString,
  OptionalAny: () => OptionalAny,
  PaymentPayloadSchema: () => PaymentPayloadSchema,
  PaymentRequiredSchema: () => PaymentRequiredSchema,
  PaymentRequirementsSchema: () => PaymentRequirementsSchema,
  ResourceInfoSchema: () => ResourceInfoSchema,
  isPaymentPayload: () => isPaymentPayload,
  isPaymentRequired: () => isPaymentRequired,
  isPaymentRequirements: () => isPaymentRequirements,
  parsePaymentPayload: () => parsePaymentPayload,
  parsePaymentRequired: () => parsePaymentRequired,
  parsePaymentRequirements: () => parsePaymentRequirements,
  validatePaymentPayload: () => validatePaymentPayload,
  validatePaymentRequired: () => validatePaymentRequired,
  validatePaymentRequirements: () => validatePaymentRequirements,
  z: () => import_zod2.z
});
module.exports = __toCommonJS(schemas_exports);
var import_zod = require("zod");
var import_zod2 = require("zod");
var NonEmptyString = import_zod.z.string().min(1);
var Any = import_zod.z.record(import_zod.z.unknown());
var OptionalAny = import_zod.z.record(import_zod.z.unknown()).optional().nullable();
var NetworkSchema = import_zod.z.string().min(3).refine((val) => val.includes(":"), {
  message: "Network must be in CAIP-2 format (e.g., 'eip155:84532')"
});
var ResourceInfoSchema = import_zod.z.object({
  url: NonEmptyString,
  description: import_zod.z.string().optional(),
  mimeType: import_zod.z.string().optional()
});
var PaymentRequirementsSchema = import_zod.z.object({
  scheme: NonEmptyString,
  network: NetworkSchema,
  amount: NonEmptyString,
  asset: NonEmptyString,
  payTo: NonEmptyString,
  maxTimeoutSeconds: import_zod.z.number().positive(),
  extra: OptionalAny
});
var PaymentRequiredSchema = import_zod.z.object({
  x402Version: import_zod.z.literal(2),
  error: import_zod.z.string().optional(),
  resource: ResourceInfoSchema,
  accepts: import_zod.z.array(PaymentRequirementsSchema).min(1),
  extensions: OptionalAny
});
var PaymentPayloadSchema = import_zod.z.object({
  x402Version: import_zod.z.literal(2),
  resource: ResourceInfoSchema.optional(),
  accepted: PaymentRequirementsSchema,
  payload: Any,
  extensions: OptionalAny
});
function parsePaymentRequired(value) {
  return PaymentRequiredSchema.safeParse(value);
}
function validatePaymentRequired(value) {
  return PaymentRequiredSchema.parse(value);
}
function isPaymentRequired(value) {
  return PaymentRequiredSchema.safeParse(value).success;
}
function parsePaymentRequirements(value) {
  return PaymentRequirementsSchema.safeParse(value);
}
function validatePaymentRequirements(value) {
  return PaymentRequirementsSchema.parse(value);
}
function isPaymentRequirements(value) {
  return PaymentRequirementsSchema.safeParse(value).success;
}
function parsePaymentPayload(value) {
  return PaymentPayloadSchema.safeParse(value);
}
function validatePaymentPayload(value) {
  return PaymentPayloadSchema.parse(value);
}
function isPaymentPayload(value) {
  return PaymentPayloadSchema.safeParse(value).success;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Any,
  NetworkSchema,
  NonEmptyString,
  OptionalAny,
  PaymentPayloadSchema,
  PaymentRequiredSchema,
  PaymentRequirementsSchema,
  ResourceInfoSchema,
  isPaymentPayload,
  isPaymentRequired,
  isPaymentRequirements,
  parsePaymentPayload,
  parsePaymentRequired,
  parsePaymentRequirements,
  validatePaymentPayload,
  validatePaymentRequired,
  validatePaymentRequirements,
  z
});
