// @okxweb3/x402-core/dist/cjs/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  OKXFacilitatorClient: () => OKXFacilitatorClient,
  x402Version: () => x402Version
});
module.exports = __toCommonJS(src_exports);

// src/facilitator/OKXFacilitatorClient.ts
var import_node_crypto = __toESM(require("crypto"));
var OKXFacilitatorClient = class {
  /**
   *
   * @param config
   */
  constructor(config) {
    this.config = {
      baseUrl: "https://web3.okx.com",
      ...config
    };
  }
  /**
   *
   * @param method
   * @param path
   * @param body
   */
  createHeaders(method, path, body) {
    const timestamp = (/* @__PURE__ */ new Date()).toISOString();
    const prehash = timestamp + method + path + (body ?? "");
    const sign = import_node_crypto.default.createHmac("sha256", this.config.secretKey).update(prehash).digest("base64");
    return {
      "OK-ACCESS-KEY": this.config.apiKey,
      "OK-ACCESS-SIGN": sign,
      "OK-ACCESS-TIMESTAMP": timestamp,
      "OK-ACCESS-PASSPHRASE": this.config.passphrase,
      "Content-Type": "application/json"
    };
  }
  /**
   *
   */
  async getSupported() {
    const path = "/api/v6/pay/x402/supported";
    const res = await fetch(this.config.baseUrl + path, {
      headers: this.createHeaders("GET", path)
    });
    if (!res.ok) throw new Error(`OKX getSupported failed: ${res.status}`);
    const json = await res.json();
    const data = json.data ?? json;
    return data;
  }
  /**
   *
   * @param payload
   * @param requirements
   */
  async verify(payload, requirements) {
    const path = "/api/v6/pay/x402/verify";
    const body = JSON.stringify({
      x402Version: 2,
      paymentPayload: payload,
      paymentRequirements: requirements
    });
    const res = await fetch(this.config.baseUrl + path, {
      method: "POST",
      headers: this.createHeaders("POST", path, body),
      body
    });
    if (!res.ok) throw new Error(`OKX verify failed: ${res.status}`);
    const json = await res.json();
    const data = json.data ?? json;
    return data;
  }
  /**
   *
   * @param payload
   * @param requirements
   */
  async settle(payload, requirements) {
    const path = "/api/v6/pay/x402/settle";
    const bodyObj = {
      x402Version: 2,
      paymentPayload: payload,
      paymentRequirements: requirements
    };
    if (this.config.syncSettle !== void 0) {
      bodyObj.syncSettle = this.config.syncSettle;
    }
    const body = JSON.stringify(bodyObj);
    const res = await fetch(this.config.baseUrl + path, {
      method: "POST",
      headers: this.createHeaders("POST", path, body),
      body
    });
    if (!res.ok) throw new Error(`OKX settle failed: ${res.status}`);
    const json = await res.json();
    const data = json.data ?? json;
    return data;
  }
  /**
   * Query on-chain settlement status by transaction hash.
   *
   * @param txHash - The transaction hash to query
   * @returns Settlement status response
   */
  async getSettleStatus(txHash) {
    const path = `/api/v6/pay/x402/settle/status?txHash=${encodeURIComponent(txHash)}`;
    const res = await fetch(this.config.baseUrl + path, {
      headers: this.createHeaders("GET", path)
    });
    if (!res.ok) throw new Error(`OKX getSettleStatus failed: ${res.status}`);
    const json = await res.json();
    const data = json.data ?? json;
    return data;
  }
};

// src/index.ts
var x402Version = 2;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  OKXFacilitatorClient,
  x402Version
});
