// @okxweb3/x402-evm/dist/esm/chunk-WNPJJN3M.mjs
import {
  authorizationTypes,
  createNonce,
  getEvmChainId
} from "./chunk-74QEGGFC.mjs";

// src/deferred/client/scheme.ts
import { getAddress } from "viem";
var AggrDeferredEvmScheme = class {
  constructor(sessionKeySigner, sessionCert) {
    this.sessionKeySigner = sessionKeySigner;
    this.sessionCert = sessionCert;
    this.scheme = "aggr_deferred";
  }
  async createPaymentPayload(x402Version, paymentRequirements, _context) {
    const nonce = createNonce();
    const now = Math.floor(Date.now() / 1e3);
    const chainId = getEvmChainId(paymentRequirements.network);
    const authorization = {
      from: this.sessionKeySigner.address,
      to: getAddress(paymentRequirements.payTo),
      value: paymentRequirements.amount,
      validAfter: (now - 5).toString(),
      validBefore: (now + paymentRequirements.maxTimeoutSeconds).toString(),
      nonce
    };
    if (!paymentRequirements.extra?.name || !paymentRequirements.extra?.version) {
      throw new Error(
        "EIP-712 domain parameters (name, version) required in payment requirements"
      );
    }
    const domain = {
      name: paymentRequirements.extra.name,
      version: paymentRequirements.extra.version,
      chainId,
      verifyingContract: getAddress(paymentRequirements.asset)
    };
    const signature = await this.sessionKeySigner.signTypedData({
      domain,
      types: authorizationTypes,
      primaryType: "TransferWithAuthorization",
      message: {
        from: getAddress(authorization.from),
        to: getAddress(authorization.to),
        value: BigInt(authorization.value),
        validAfter: BigInt(authorization.validAfter),
        validBefore: BigInt(authorization.validBefore),
        nonce: authorization.nonce
      }
    });
    return {
      x402Version,
      payload: { authorization, signature },
      extensions: {
        scheme: "aggr_deferred"
      },
      acceptedExtraOverrides: {
        sessionCert: this.sessionCert
      }
    };
  }
};

export {
  AggrDeferredEvmScheme
};
