// @okxweb3/x402-evm/dist/esm/chunk-WKFT4EOY.mjs
import {
  getDefaultAsset
} from "./chunk-A7UQ7D5C.mjs";

// src/deferred/server/scheme.ts
var AggrDeferredEvmScheme = class {
  constructor() {
    this.scheme = "aggr_deferred";
  }
  async parsePrice(price, network) {
    if (typeof price === "object" && price !== null && "amount" in price) {
      if (!price.asset) {
        throw new Error(`Asset address must be specified for AssetAmount on network ${network}`);
      }
      return {
        amount: price.amount,
        asset: price.asset,
        extra: price.extra || {}
      };
    }
    const amount = this.parseMoneyToDecimal(price);
    return this.defaultMoneyConversion(amount, network);
  }
  enhancePaymentRequirements(paymentRequirements, supportedKind, extensionKeys) {
    void supportedKind;
    void extensionKeys;
    return Promise.resolve(paymentRequirements);
  }
  parseMoneyToDecimal(money) {
    if (typeof money === "number") {
      return money;
    }
    const cleanMoney = money.replace(/^\$/, "").trim();
    const amount = parseFloat(cleanMoney);
    if (isNaN(amount)) {
      throw new Error(`Invalid money format: ${money}`);
    }
    return amount;
  }
  defaultMoneyConversion(amount, network) {
    const assetInfo = getDefaultAsset(network);
    const tokenAmount = this.convertToTokenAmount(amount.toString(), assetInfo.decimals);
    const includeEip712Domain = !assetInfo.assetTransferMethod || assetInfo.supportsEip2612;
    return {
      amount: tokenAmount,
      asset: assetInfo.address,
      extra: {
        ...includeEip712Domain && {
          name: assetInfo.name,
          version: assetInfo.version
        },
        ...assetInfo.assetTransferMethod && {
          assetTransferMethod: assetInfo.assetTransferMethod
        }
      }
    };
  }
  convertToTokenAmount(decimalAmount, decimals) {
    const amount = parseFloat(decimalAmount);
    if (isNaN(amount)) {
      throw new Error(`Invalid amount: ${decimalAmount}`);
    }
    const [intPart, decPart = ""] = String(amount).split(".");
    const paddedDec = decPart.padEnd(decimals, "0").slice(0, decimals);
    const tokenAmount = (intPart + paddedDec).replace(/^0+/, "") || "0";
    return tokenAmount;
  }
};

export {
  AggrDeferredEvmScheme
};
