// @okxweb3/x402-evm/dist/esm/index.mjs
import {
  AggrDeferredEvmScheme as AggrDeferredEvmScheme2
} from "./chunk-WKFT4EOY.mjs";
import {
  ExactEvmScheme
} from "./chunk-2HIS7LN3.mjs";
import {
  UptoEvmScheme
} from "./chunk-HEMO754C.mjs";
import {
  createPermit2ApprovalTx,
  getPermit2AllowanceReadParams
} from "./chunk-QSDGM24Y.mjs";
import "./chunk-A7UQ7D5C.mjs";
import {
  isEIP3009Payload,
  isPermit2Payload,
  isUptoPermit2Payload
} from "./chunk-KCWCNQAZ.mjs";
import "./chunk-2K22OMTA.mjs";
import {
  AggrDeferredEvmScheme
} from "./chunk-WNPJJN3M.mjs";
import {
  PERMIT2_ADDRESS,
  authorizationTypes,
  eip3009ABI,
  erc20AllowanceAbi,
  permit2WitnessTypes,
  uptoPermit2WitnessTypes,
  x402ExactPermit2ProxyABI,
  x402ExactPermit2ProxyAddress,
  x402UptoPermit2ProxyABI,
  x402UptoPermit2ProxyAddress
} from "./chunk-74QEGGFC.mjs";

// src/signer.ts
function toClientEvmSigner(signer, publicClient) {
  const readContract = signer.readContract ?? publicClient?.readContract.bind(publicClient);
  const result = {
    address: signer.address,
    signTypedData: (msg) => signer.signTypedData(msg)
  };
  if (readContract) {
    result.readContract = readContract;
  }
  const signTransaction = signer.signTransaction;
  if (signTransaction) {
    result.signTransaction = (args) => signTransaction(args);
  }
  const getTransactionCount = signer.getTransactionCount ?? publicClient?.getTransactionCount?.bind(publicClient);
  if (getTransactionCount) {
    result.getTransactionCount = (args) => getTransactionCount(args);
  }
  const estimateFeesPerGas = signer.estimateFeesPerGas ?? publicClient?.estimateFeesPerGas?.bind(publicClient);
  if (estimateFeesPerGas) {
    result.estimateFeesPerGas = () => estimateFeesPerGas();
  }
  return result;
}
function toFacilitatorEvmSigner(client) {
  return {
    ...client,
    getAddresses: () => [client.address]
  };
}
export {
  AggrDeferredEvmScheme as AggrDeferredEvmClientScheme,
  AggrDeferredEvmScheme2 as AggrDeferredEvmServerScheme,
  ExactEvmScheme,
  PERMIT2_ADDRESS,
  UptoEvmScheme,
  authorizationTypes,
  createPermit2ApprovalTx,
  eip3009ABI,
  erc20AllowanceAbi,
  getPermit2AllowanceReadParams,
  isEIP3009Payload,
  isPermit2Payload,
  isUptoPermit2Payload,
  permit2WitnessTypes,
  toClientEvmSigner,
  toFacilitatorEvmSigner,
  uptoPermit2WitnessTypes,
  x402ExactPermit2ProxyABI,
  x402ExactPermit2ProxyAddress,
  x402UptoPermit2ProxyABI,
  x402UptoPermit2ProxyAddress
};
