// @okxweb3/x402-evm/dist/cjs/deferred/server/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/deferred/server/index.ts
var server_exports = {};
__export(server_exports, {
  AggrDeferredEvmScheme: () => AggrDeferredEvmScheme
});
module.exports = __toCommonJS(server_exports);

// src/shared/defaultAssets.ts
var DEFAULT_STABLECOINS = {
  "eip155:196": {
    address: "0x779ded0c9e1022225f8e0630b35a9b54be713736",
    name: "USD\u20AE0",
    version: "1",
    decimals: 6
  },
  // X Layer mainnet USDT0 (EIP-3009)
  "eip155:1952": {
    address: "0x9e29b3aada05bf2d2c827af80bd28dc0b9b4fb0c",
    name: "USD\u20AE0",
    version: "1",
    decimals: 6
  }
};
function getDefaultAsset(network) {
  const info = DEFAULT_STABLECOINS[network];
  if (!info) {
    throw new Error(`No default asset configured for network ${network}`);
  }
  return info;
}

// src/deferred/server/scheme.ts
var AggrDeferredEvmScheme = class {
  constructor() {
    this.scheme = "aggr_deferred";
  }
  async parsePrice(price, network) {
    if (typeof price === "object" && price !== null && "amount" in price) {
      if (!price.asset) {
        throw new Error(`Asset address must be specified for AssetAmount on network ${network}`);
      }
      return {
        amount: price.amount,
        asset: price.asset,
        extra: price.extra || {}
      };
    }
    const amount = this.parseMoneyToDecimal(price);
    return this.defaultMoneyConversion(amount, network);
  }
  enhancePaymentRequirements(paymentRequirements, supportedKind, extensionKeys) {
    void supportedKind;
    void extensionKeys;
    return Promise.resolve(paymentRequirements);
  }
  parseMoneyToDecimal(money) {
    if (typeof money === "number") {
      return money;
    }
    const cleanMoney = money.replace(/^\$/, "").trim();
    const amount = parseFloat(cleanMoney);
    if (isNaN(amount)) {
      throw new Error(`Invalid money format: ${money}`);
    }
    return amount;
  }
  defaultMoneyConversion(amount, network) {
    const assetInfo = getDefaultAsset(network);
    const tokenAmount = this.convertToTokenAmount(amount.toString(), assetInfo.decimals);
    const includeEip712Domain = !assetInfo.assetTransferMethod || assetInfo.supportsEip2612;
    return {
      amount: tokenAmount,
      asset: assetInfo.address,
      extra: {
        ...includeEip712Domain && {
          name: assetInfo.name,
          version: assetInfo.version
        },
        ...assetInfo.assetTransferMethod && {
          assetTransferMethod: assetInfo.assetTransferMethod
        }
      }
    };
  }
  convertToTokenAmount(decimalAmount, decimals) {
    const amount = parseFloat(decimalAmount);
    if (isNaN(amount)) {
      throw new Error(`Invalid amount: ${decimalAmount}`);
    }
    const [intPart, decPart = ""] = String(amount).split(".");
    const paddedDec = decPart.padEnd(decimals, "0").slice(0, decimals);
    const tokenAmount = (intPart + paddedDec).replace(/^0+/, "") || "0";
    return tokenAmount;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AggrDeferredEvmScheme
});
