// @okxweb3/x402-evm/dist/cjs/deferred/client/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/deferred/client/index.ts
var client_exports = {};
__export(client_exports, {
  AggrDeferredEvmScheme: () => AggrDeferredEvmScheme
});
module.exports = __toCommonJS(client_exports);

// src/constants.ts
var authorizationTypes = {
  TransferWithAuthorization: [
    { name: "from", type: "address" },
    { name: "to", type: "address" },
    { name: "value", type: "uint256" },
    { name: "validAfter", type: "uint256" },
    { name: "validBefore", type: "uint256" },
    { name: "nonce", type: "bytes32" }
  ]
};

// src/utils.ts
var import_viem = require("viem");
function getEvmChainId(network) {
  if (network.startsWith("eip155:")) {
    const idStr = network.split(":")[1];
    const chainId = parseInt(idStr, 10);
    if (isNaN(chainId)) {
      throw new Error(`Invalid CAIP-2 chain ID: ${network}`);
    }
    return chainId;
  }
  throw new Error(`Unsupported network format: ${network} (expected eip155:CHAIN_ID)`);
}
function getCrypto() {
  const cryptoObj = globalThis.crypto;
  if (!cryptoObj) {
    throw new Error("Crypto API not available");
  }
  return cryptoObj;
}
function createNonce() {
  return (0, import_viem.toHex)(getCrypto().getRandomValues(new Uint8Array(32)));
}

// src/deferred/client/scheme.ts
var import_viem2 = require("viem");
var AggrDeferredEvmScheme = class {
  constructor(sessionKeySigner, sessionCert) {
    this.sessionKeySigner = sessionKeySigner;
    this.sessionCert = sessionCert;
    this.scheme = "aggr_deferred";
  }
  async createPaymentPayload(x402Version, paymentRequirements, _context) {
    const nonce = createNonce();
    const now = Math.floor(Date.now() / 1e3);
    const chainId = getEvmChainId(paymentRequirements.network);
    const authorization = {
      from: this.sessionKeySigner.address,
      to: (0, import_viem2.getAddress)(paymentRequirements.payTo),
      value: paymentRequirements.amount,
      validAfter: (now - 5).toString(),
      validBefore: (now + paymentRequirements.maxTimeoutSeconds).toString(),
      nonce
    };
    if (!paymentRequirements.extra?.name || !paymentRequirements.extra?.version) {
      throw new Error(
        "EIP-712 domain parameters (name, version) required in payment requirements"
      );
    }
    const domain = {
      name: paymentRequirements.extra.name,
      version: paymentRequirements.extra.version,
      chainId,
      verifyingContract: (0, import_viem2.getAddress)(paymentRequirements.asset)
    };
    const signature = await this.sessionKeySigner.signTypedData({
      domain,
      types: authorizationTypes,
      primaryType: "TransferWithAuthorization",
      message: {
        from: (0, import_viem2.getAddress)(authorization.from),
        to: (0, import_viem2.getAddress)(authorization.to),
        value: BigInt(authorization.value),
        validAfter: BigInt(authorization.validAfter),
        validBefore: BigInt(authorization.validBefore),
        nonce: authorization.nonce
      }
    });
    return {
      x402Version,
      payload: { authorization, signature },
      extensions: {
        scheme: "aggr_deferred"
      },
      acceptedExtraOverrides: {
        sessionCert: this.sessionCert
      }
    };
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AggrDeferredEvmScheme
});
